### PR TITLE
ANSIComponentSerializer

### LINKED ISSUE #7: Adventure component implementation
There should be an implementation of ansi that works with adventure components

I'm not sure if that's better suited to be here or in the adventure repo -- it wouldn't fully implement the deserialization side of the ComponentSerializer interface, at least at first.

### LINKED ISSUE #125: [Feature Request] Add a serializer that produces ansi formatted text including colors
The ansi console can support up to full true colour.  That being said it is very hard to detect the level of color available in the console.

the Jansi library is one option there might be others.

The goal is here is to create a formatted text object that can be sent to the console and will display consistently across as many viewers as possible.

### PR DESCRIPTION
**Pull request by rymiel**
Supersedes #140
Fixes KyoriPowered/ansi#7
Fixes #125

### CODE REVIEW

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * A serializer which emits <a href="https://en.wikipedia.org/wiki/ANSI_escape_code">ANSI escape sequences</a>.
+ * Note that this serializer does NOT support serialization.
+ *
+ * @since 4.14.0
+ */
+public interface ANSIComponentSerializer extends ComponentSerializer<Component, Component, String> {
+  /**
+   * Gets a component serializer for serialization to a string using ANSI escape codes.
+   * Note that this serializer does NOT support serialization.
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]:
```suggestion
   * 
   * <p>Note that this serializer does NOT support deserialization.</p>
```

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A serializer which emits <a href="https://en.wikipedia.org/wiki/ANSI_escape_code">ANSI escape sequences</a>.
+ * Note that this serializer does NOT support serialization.
+ *
+ * @since 4.14.0
+ */
+public interface ANSIComponentSerializer extends ComponentSerializer<Component, Component, String> {
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]: We should probably add a builder that allows configuring:

- default color level (autodetect or sth fixed)
- component flattener used
- i think that's it?

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Serialize a {@link Component} to a {@link String} with ANSI escape sequences at the given {@link ColorLevel}.
+   *
+   * @param component the component to serialize
+   * @param colorLevel the color level to use for escape sequences
+   * @return string with escape sequences
+   * @see ColorLevel
+   * @since 4.14.0
+   */
+  @NotNull String serialize(@NotNull Component component, @NotNull ColorLevel colorLevel);
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]: I don't know if we even want to expose this as an overload? like i think if the color level (or colour level if you want to have fun) being settable in the builder for a serializer instance should be fine?

**Review by zml2008 — CHANGES_REQUESTED** [review]
can't rly words rn but some initial thoughts

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java (hunk 3 above)
Comment (rymiel) [reply text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]: yep, i only added it cuz i was too lazy to make a whole builder, but it exists now

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Sets the component flattener instance to use when traversing the component for serialization.
+     *
+     * <p>By default, this serializer will use {@link ComponentFlattener#basic()}.</p>
+     *
+     * @param componentFlattener the flattener instance.
+     * @return this builder
+     * @since 4.14.0
+     */
+    @NotNull Builder flattener(final ComponentFlattener componentFlattener);
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]:
```suggestion
    @NotNull Builder flattener(final @NotNull ComponentFlattener componentFlattener);
```

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+     * Sets the default color level used when serializing.
+     *
+     * <p>By default, this serializer will use {@link ColorLevel#compute()} to try to detect the color level of the terminal being used.</p>
+     *
+     * @param colorLevel the color level
+     * @return this builder
+     * @see ColorLevel
+     * @since 4.14.0
+     */
+    @NotNull Builder colorLevel(final ColorLevel colorLevel);
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]:
```suggestion
    @NotNull Builder colorLevel(final @NotNull ColorLevel colorLevel);
```

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  static final class BuilderImpl implements ANSIComponentSerializer.Builder {
+    private ColorLevel colorLevel = ColorLevel.compute();
+    private ComponentFlattener flattener = ComponentFlattener.basic();
+
+    BuilderImpl() {
+      BUILDER.accept(this);
+    }
+
+    @Override
+    public @NotNull Builder colorLevel(final ColorLevel colorLevel) {
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java]:
```suggestion
    public @NotNull Builder colorLevel(final @NotNull ColorLevel colorLevel) {
```

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+      this.colorLevel = colorLevel;
+      return this;
+    }
+
+    @Override
+    public @NotNull Builder flattener(final ComponentFlattener componentFlattener) {
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java]:
```suggestion
    public @NotNull Builder flattener(final @NotNull ComponentFlattener componentFlattener) {
```

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+ * A serializer which emits <a href="https://en.wikipedia.org/wiki/ANSI_escape_code">ANSI escape sequences</a>.
+ * Note that this serializer does NOT support serialization.
+ *
+ * @since 4.14.0
+ */
+public interface ANSIComponentSerializer extends ComponentSerializer<Component, Component, String> {
+  /**
+   * Gets a component serializer for serialization to a string using ANSI escape codes.
+   *
+   * <p>Note that this serializer does NOT support serialization.</p>
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]:
```suggestion
   * <p>Note that this serializer does NOT support deserialization.</p>
```

**Review by zml2008 — CHANGES_REQUESTED** [review]
This is pretty close, just a few nits here!

We should see about introducing an extra interface for these sorts of serializers that only defines the `serialize()` method -- maybe `ComponentEncoder`? or something? We've already used up serializer and renderer lol

**Review by zml2008 — APPROVED** [review]
Looks good to merge as soon as you bump the `ansi` dependency to the release version :)

File: text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+import net.kyori.adventure.text.flattener.ComponentFlattener;
+import net.kyori.adventure.text.serializer.ComponentEncoder;
+import net.kyori.adventure.util.PlatformAPI;
+import net.kyori.ansi.ColorLevel;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A serializer which emits <a href="https://en.wikipedia.org/wiki/ANSI_escape_code">ANSI escape sequences</a>.
+ * Note that this serializer does NOT support serialization.
```
Comment (zml2008) [inline text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializer.java]:
```suggestion
 *
 * <p>Note that this serializer does NOT support serialization.</p>
```

**Review by zml2008 — CHANGES_REQUESTED** [review]
ok one last thing sorry :(

**Review by zml2008 — APPROVED** [review]

**Pull request merged**